### DISCUSSION TITLE
Guys I need help. Do I need to use malloc???

### DISCUSSION THREAD

**Discussion by jeff-221** [community/community#59889] [Programming Help]
Hello, 
I have to submit a project for the university. My task is to create a C-library for Matrix-Vector calculations. Im almost finished, but now I ask myself if I need to use the malloc function every time I create a new struct?? Thanks.
For example: 

typedef struct{
double mat[3][3];
}matrix;

int main(){
......
}

matrix inverse(matrix * m){
    double det = determinante(m);

    if(det == 0){
        printf("Es gibt keine Inverse\n");
        exit(1); 
        
    }
    else{
        double  KehrwDet = 1 / det; 
        matrix invers;        //malloc ?????? do I need to use the malloc fuction
        for(int i = 0; i < 3; i++){
            for(int k = 0; k < 3; k++){
                invers.mat[i][k] = ( KehrwDet * (m->mat[(k+1)%3][(i+1)%3] * m->mat[(k+2)%3][(i+2)%3] - m->mat[(k+1)%3][(i+2)%3] * m->mat[(k+2)%3][(i+1)%3])) ; 
                printf(" %f ", invers.mat[i][k]); 
            }
            printf("\n");
        }
        return invers;

    }

}

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by leonden** [comment]
In the code you provided, you have declared the `matrix` struct as a stack variable inside the `inverse` function. In this case, you don't need to use `malloc` to allocate memory for the struct.

Using `malloc` is necessary when you want to allocate memory on the heap (dynamically) instead of the stack (statically). If you use `malloc`, you would allocate memory like this:

```c
matrix *invers = malloc(sizeof(matrix));
```

Then, you would need to free the allocated memory using `free(invers);` when you no longer need it.

However, in your case, since you have declared `matrix invers;` as a local variable inside the function, it is allocated on the stack automatically. Therefore, there's no need to use `malloc` in this scenario.

One thing to note is that when you return a struct from a function, you need to ensure that the struct's memory is still valid when accessed outside the function. In your current code, when the `inverse` function returns, the local `invers` variable will go out of scope, and accessing it outside the function will lead to undefined behavior.

To address this issue, you have a few options:

1. Pass a pointer to the `invers` struct as a parameter to the function and modify it directly inside the function.
2. Allocate memory dynamically using `malloc` inside the function and return a pointer to the allocated memory. In this case, the caller of the function would be responsible for freeing the memory when it's no longer needed.
3. Declare the `invers` struct as a static variable inside the function. However, note that using a static variable may lead to issues in multithreaded environments or when calling the function multiple times.

Choose the option that best suits your requirements and design.

**Comment by davea42** [comment]
This C (the poster says) so   'return invers;' 
returns a full copy of the struct.  So callers can access the returned struct..
The invers *inside* the function will indeed go away, but such
is irrelevant as what is returned is not a pointer, it is a  copy of the struct.

If the struct were to be large it would be somewhat unusual to
return (meaning copy) a struct.

**Comment by RamtinKosari** [comment]
Hi dear @jeff-221 
In your code, you are declaring the invers variable of type matrix as a local variable within the inverse function. In this case, you do not need to use malloc to allocate memory for it because it is a stack-allocated variable.

When you declare a variable in this way, memory is automatically allocated on the stack when the function is called, and it is automatically deallocated when the function returns. You can safely return the invers variable as it will still be valid after the function returns.

So, in your code, there is no need to use malloc for the invers variable. The stack allocation is sufficient for your purposes.

However, if you were dynamically allocating memory for a struct using malloc or a similar function, you would need to free that memory explicitly using free to avoid memory leaks. But in this case, since you are using a stack-allocated variable, you don't need to worry about freeing the memory explicitly.

Here's an example code that demonstrates the usage of a stack-allocated variable without the need for malloc:
```c++
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    double mat[3][3];
} matrix;

matrix inverse(matrix *m) {
    double det = determinante(m);
    if (det == 0) {
        printf("Es gibt keine Inverse\n");
        exit(1);
    } else {
        double KehrwDet = 1 / det;
        matrix invers;
        for (int i = 0; i < 3; i++) {
            for (int k = 0; k < 3; k++) {
                invers.mat[i][k] = (KehrwDet * (m->mat[(k + 1) % 3][(i + 1) % 3] * m->mat[(k + 2) % 3][(i + 2) % 3] - m->mat[(k + 1) % 3][(i + 2) % 3] * m->mat[(k + 2) % 3][(i + 1) % 3]));
                printf(" %f ", invers.mat[i][k]);
            }
            printf("\n");
        }
        return invers;
    }
}

int main() {
    matrix m;
    // Initialize the matrix m with values
    
    matrix inv = inverse(&m);
    
    // Use the inverse matrix inv
    
    return 0;
}
```
In this example, the invers matrix is declared as a local variable within the inverse function. It is stack-allocated and does not require dynamic memory allocation (malloc). The inverse function calculates the inverse of the input matrix m and returns the resulting inverse matrix.

In the main function, you can create a matrix m and pass its address (&m) to the inverse function. The function will calculate the inverse and return it as a stack-allocated matrix struct.

You can then use the inverse matrix inv as needed within the main function.

Remember to replace the determinante function with your actual implementation of calculating the determinant of a matrix.
